File: src/main/java/com/jvanev/jxconfig/annotation/DependsOn.java
/*
 * Copyright 2025 Georgi Vanev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jvanev.jxconfig.annotation;

import com.jvanev.jxconfig.resolver.DependencyChecker;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks a {@link ConfigProperty} or a {@link ConfigGroup} as dependent on a specified {@link ConfigProperty}'s value.
 */
@Target({ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
public @interface DependsOn {
    /**
     * The {@link ConfigProperty#name()} of the property that will determine the runtime value of this parameter.
     *
     * @return The {@link ConfigProperty#name()} of the dependency.
     */
    String property();

    /**
     * The operator to be used for condition check.
     * <p>
     * Defaults to {@link DependencyChecker#DEFAULT_OPERATOR}, meaning the default,
     * case-sensitive string comparison is used.
     * <p>
     * <b>Warning:</b> If no custom {@link DependencyChecker} has been registered,
     * a {@code NullPointerException} will be thrown when performing the check.
     *
     * @return The condition check operator.
     */
    String operator() default DependencyChecker.DEFAULT_OPERATOR;

    /**
     * The value the dependency must have in order for the dependency condition to be satisfied.
     * <p>
     * Defaults to {@code true} to emulate boolean-style comparison.
     *
     * @return The value satisfying the dependency condition.
     */
    String value() default "true";
}

File: src/main/kotlin/annotation/DependsOn.kt
/*
 * Copyright 2025 Georgi Vanev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jvanev.kconfig.annotation

/**
 * Marks a *Configuration Property* as conditionally resolved based on another property's value.
 *
 * This annotation is applied to a *Configuration Property* to specify that its value at runtime
 * depends on the resolved value of a declared *Dependency*.
 *
 * ## Terms
 *
 * - **Configuration Property:** A property declared in the primary constructor of a
 * *Configuration Type*, annotated with [ConfigProperty].
 *
 * - **Dependency:** Another *Configuration Property* within the same *Configuration Type*
 * whose [ConfigProperty.name] matches the [property] value.
 *
 * ## How it Works
 *
 * If the *Dependency*'s resolved value matches the [value] specified for *this* property,
 * *this* property's runtime value will be fetched from the `.properties` file.
 * Otherwise, its [ConfigProperty.defaultValue] will be used.
 *
 * All comparisons for dependency conditions are case-sensitive.
 *
 * @property property The name of the *Dependency* property.
 * @property value The expected value of the *Dependency* property for *this* configuration property to be active.
 * Defaults to `True` for boolean-style activation.
 */
@MustBeDocumented
@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class DependsOn(val property: String, val value: String = "True")
